### ISSUE TITLE
Fix python 2.7.9+ certificate check for S3-compatible services

### ISSUE THREAD

**Issue by jamshid** [s3tools/s3cmd#612]
This is related to https://github.com/s3tools/s3cmd/issues/437, hoping you can provide some direction on what needs to be done to get s3cmd working with S3-compatible a https endpoint like https://me.cloud.example.com.

I'm using master but it still doesn't work on my OSX with python 2.7.10.

```
$ git clone https://github.com/s3tools/s3cmd.git
$ cd s3cmd
$ python setup.py build

# Create a file containing each cert, hand edited to remove non-cert output
$ openssl s_client -connect me.cloud.example.com:443 -showcerts </dev/null 2>/dev/null > mycertfile.pem

# List buckets works
$ ./build/scripts-2.7/s3cmd --no-check-certificate --ca-cert ./mycertfile.pem ls 
2015-02-05 07:51  s3://x
2015-02-04 04:15  s3://yum

# List the "x" bucket, weird that it works, right?! I think it's related to being a single character.
$ ./build/scripts-2.7/s3cmd -d --no-check-certificate --ca-cert ./mycertfile.pem ls s3://x
DEBUG: Using ca_certs_file ./mycertfile.pem
DEBUG: Disabling hostname checking
DEBUG: non-proxied HTTPSConnection(me.cloud.example.com)
DEBUG: Unicodising "Bucket name 'x' is too short (min 3 characters)" using UTF-8
DEBUG: format_uri(): /x/?delimiter=/
...
                       DIR   s3://x/repodata/
...

# Listing any other bucket fails :-(
$ ./build/scripts-2.7/s3cmd -d --no-check-certificate --ca-cert ./mycertfile.pem ls s3://yum
WARNING: Retrying failed request: /?delimiter=/ (hostname 'yum.me.cloud.example.com' doesn't match either of '*.cloud.example.com', 'cloud.example.com')
...
```

This is my ~/.s3cfg:
```
[default]
access_key = XXX
host_base = me.cloud.example.com
host_bucket = %(bucket)s.me.cloud.example.com
secret_key = secret
signature_v2=True
use_https = True
```

**Comment by jamshid** [comment]
It's not a general solution, but a side-effect of always returning False from check_bucket_name() is that it avoids the certificate check failure. If you have any pointers maybe I can come up with a usable pull request.

**Comment by mdomsch** [comment]
Please try with this:


diff --git a/S3/ConnMan.py b/S3/ConnMan.py
index 65b1050..c311efe 100644
--- a/S3/ConnMan.py
+++ b/S3/ConnMan.py
@@ -159,7 +159,7 @@ class ConnMan(object):
             debug("ConnMan.get(): creating new connection: %s" % conn_id)
             conn = http_connection(conn_id, hostname, ssl, cfg)
             conn.c.connect()
-            if conn.ssl:
+            if conn.ssl and cfg.check_ssl_certificate:
                 conn.match_hostname()
         conn.counter += 1
         return conn






On Wed, Sep 2, 2015 at 10:35 PM, jamshid <notifications@github.com> wrote:
>
> It's not a general solution, but a side-effect of always returning False
from check_bucket_name() is that it avoids the certificate check failure.
If you have any pointers maybe I can come up with a usable pull request.

**Comment by mdomsch** [comment]
This is actually better and correct:

diff --git a/S3/ConnMan.py b/S3/ConnMan.py
index 65b1050..04a6363 100644
--- a/S3/ConnMan.py
+++ b/S3/ConnMan.py
@@ -159,7 +159,7 @@ class ConnMan(object):
             debug("ConnMan.get(): creating new connection: %s" % conn_id)
             conn = http_connection(conn_id, hostname, ssl, cfg)
             conn.c.connect()
-            if conn.ssl:
+            if conn.ssl and conn.context.check_hostname:
                 conn.match_hostname()
         conn.counter += 1
         return conn

**Comment by mdomsch** [comment]
Nope - I was right with the first patch, checking cfg.check_ssl_certificate
before calling conn.match_hostname(). Otherwise we disable checking for all
Amazon buckets (not what we want).

**Comment by fviard** [comment]
@jamshid  Is it possible for you to confirm that the fix in master is now working for you? so we can close this issue. Thanks

**Comment by jamshid** [comment]
Confirmed, the fix works. If I use the released 1.5.2 I get the failure. If I instead use latest `master` it works. Btw it would be good if `master` reported a different version than 1.5.2.
```
$ cat ~/.s3cfg
[default]
access_key = XXX
host_base = mybucket.cloud.example.com
host_bucket = %(bucket)s.mybucket.cloud.example.com
secret_key = secret
signature_v2=True
check_ssl_certificate = False
use_https = True
$ s3cmd --version
s3cmd version 1.5.2
$ s3cmd ls s3://yum
WARNING: Retrying failed request: /?delimiter=/ (hostname 'yum.mybucket.cloud.example.com' doesn't match either of '*.cloud.example.com', 'cloud.example.com')
WARNING: Waiting 3 sec...
...
$ brew uninstall s3cmd
$ brew install --HEAD s3cmd
...
==> Cloning https://github.com/s3tools/s3cmd.git
==> Checking out branch master
...
$ s3cmd --version
s3cmd version 1.5.2      # <== confusing
$ s3cmd ls s3://yum
                       DIR   s3://yum/.repodata/
                       DIR   s3://yum/repodata/
                       ...
```
